### ISSUE TITLE
Crashes due to layout modification from background thread

### ISSUE THREAD

**Issue by tiwoc** [helpscout/beacon-ios-sdk#157]
In `BeaconProgressView.swift:34` and `ChatViewModel.swift:76`, there are instances where the auto layout engine is modified from a background thread, which causes crashes.

**Technical Details:**

- Beacon install method: CocoaPods
- SDK version: 2.2.0
- iOS OS version: 14.4.1
- Device make/model: iPhone 11 Pro
- Impacted user count: 2

**Actual outcome**

The app crashes with this message:
> NSInternalInconsistencyException: Modifications to the layout engine must not be performed from a background thread after it has been accessed from the main thread.

**Expected outcome**

Beacon shouldn't cause crashes.

**Steps to reproduce**

Unknown

**Additional Information**

`BeaconProgressView.swift` crash log:

```
Hardware Model:     iPhone12,3
Role:               Foreground
OS Version:         iOS 14.4.1


NSInternalInconsistencyException: Modifications to the layout engine must not be performed from a background thread after it has been accessed from the main thread.

0  CoreFoundation          ___exceptionPreprocess
1  libobjc.A.dylib         _objc_exception_throw
2  CoreAutoLayout          __AssertAutoLayoutOnAllowedThreadsOnly
3  CoreAutoLayout          -[NSISEngine withBehaviors:performModifications:]
4  UIKitCore               -[UIView(Hierarchy) _postMovedFromSuperview:]
5  UIKitCore               ___UIViewWasRemovedFromSuperview
6  UIKitCore               -[UIView(Hierarchy) removeFromSuperview]
7  <redacted app name>     observeValue (BeaconProgressView.swift:34:18)
8  <compiler-generated>    observeValue
9  Foundation              _NSKeyValueNotifyObserver
10 Foundation              _NSKeyValueDidChange
11 Foundation              _NSKeyValueDidChangeWithPerThreadPendingNotifications
12 CoreFoundation          ___NSARRAY_IS_CALLING_OUT_TO_A_BLOCK__
13 CoreFoundation          -[__NSArrayM enumerateObjectsWithOptions:usingBlock:]
14 Foundation              -[NSProgress _setValueForKeys:settingBlock:]
15 Foundation              -[NSProgress _updateChild:fraction:portion:]
16 Foundation              -[NSProgress _updateFractionCompleted:]
17 Foundation              -[NSProgress setCompletedUnitCount:]
18 Foundation              ___88-[NSXPCConnection _sendInvocation:orArguments:count:methodSignature:selector:withProxy:]_block_invoke_3
19 libxpc.dylib            __xpc_connection_reply_callout
20 libxpc.dylib            __xpc_connection_call_reply_async
21 libdispatch.dylib       __dispatch_client_callout3
22 libdispatch.dylib       __dispatch_mach_msg_async_reply_invoke
23 libdispatch.dylib       __dispatch_lane_serial_drain
24 libdispatch.dylib       __dispatch_lane_invoke
25 libdispatch.dylib       __dispatch_workloop_worker_thread
26 libsystem_pthread.dylib __pthread_wqthread
27 libsystem_pthread.dylib _start_wqthread
```

`ChatViewModel.swift` crash log:

```
Hardware Model:     iPhone12,3
Role:               Foreground
OS Version:         iOS 14.4.1


NSInternalInconsistencyException: Modifications to the layout engine must not be performed from a background thread after it has been accessed from the main thread.

0  CoreFoundation          ___exceptionPreprocess
1  libobjc.A.dylib         _objc_exception_throw
2  CoreAutoLayout          __AssertAutoLayoutOnAllowedThreadsOnly
3  CoreAutoLayout          -[NSISEngine removeConstraintWithMarker:]
4  CoreAutoLayout          -[NSLayoutConstraint _removeFromEngine:]
5  UIKitCore               -[UIView(UIConstraintBasedLayout) _layoutEngine_willRemoveLayoutConstraint:]
6  UIKitCore               -[UIView(UIConstraintBasedLayout) _constraints_willChangeAutoresizingConstraintsArrayForContainedView:]
7  UIKitCore               -[UIView(UIConstraintBasedLayout) _setAutoresizingConstraints:]
8  UIKitCore               -[UIView(UIConstraintBasedLayout) _invalidateAutoresizingConstraints]
9  UIKitCore               -[UIView(UIConstraintBasedLayout) _constraints_subviewWillChangeSuperview:]
10 UIKitCore               ___UIViewWillBeRemovedFromSuperview
11 UIKitCore               -[UIView(Hierarchy) removeFromSuperview]
12 UIKitCore               -[UICollectionView _reuseCell:notifyDidEndDisplaying:]
13 CoreFoundation          ___NSDICTIONARY_IS_CALLING_OUT_TO_A_BLOCK__
14 CoreFoundation          -[__NSDictionaryM enumerateKeysAndObjectsWithOptions:usingBlock:]
15 UIKitCore               ___30-[UICollectionView reloadData]_block_invoke
16 UIKitCore               -[UICollectionView reloadData]
17 <redacted app name>     process (ChatViewModel.swift:76:28)
18 <redacted app name>     partial apply forwarder for closure #1 (Beacon.ChatState) -> () in Beacon.ChatViewModel.init(collectionView: __C.UICollectionView, inputBar: Beacon.ChatInputBar, store: Beacon.Store<Beacon.ChatState, Beacon.ChatViewAction>, eventCountsTimerDuration: Swift.Double, scrollHandler: (__C.UIScrollView, __C.CGPoint, __C.CGPoint) -> (), scrollCompletionObserver: () -> ()) -> Beacon.ChatViewModel (ChatViewModel.swift:43:19)
19 <compiler-generated>    partial apply forwarder for generic specialization <Beacon.ChatState, Beacon.ChatViewAction> of reabstraction thunk helper <A, B> from @escaping @callee_guaranteed (@in_guaranteed A) -> () to @escaping @callee_guaranteed (@in_guaranteed A) -> (@out ())
20 <compiler-generated>    generic specialization <Beacon.ChatState, Beacon.ChatViewAction> of closure #1 (B) -> () in closure #1 (Beacon.Effect<B>) -> () in Beacon.Store.reduce(B) -> [Beacon.Effect<B>]
21 <redacted app name>     generic specialization <Swift.Result<Foundation.Data, Beacon.ChatAPI.ChatError>, Beacon.ChatViewAction> of closure #1 (A) -> () in closure #1 ((A1) -> ()) -> () in Beacon.Effect.map<A>((A) -> A1) -> Beacon.Effect<A1> (Effect.swift:11:52)
22 <redacted app name>     partial apply forwarder for generic specialization <Beacon.ChatAPI.ChatError, Foundation.Data> of closure #1 (Swift.Result<A1, A>) -> () in closure #1 ((Swift.Result<A1, A>) -> ()) -> () in Beacon.APIClient.load<A where A1: Swift.Decodable>(Beacon.RemoteEndpoint<A1>) -> Beacon.Effect<Swift.Result<A1, A>> (APIClient.swift:105:17)
23 <redacted app name>     handle (APIClient.swift:258:9)
24 Beacon                  merged partial apply forwarder for generic specialization <Beacon.ChatAPI.ChatError, [Beacon.APIPusherChannel]> of closure #2 (Foundation.Data?, __C.NSURLResponse?, Swift.Error?) -> () in Beacon.APIClient.load<A where A1: Swift.Decodable>(_: Beacon.RemoteEndpoint<A1>, completion: (Swift.Result<A1, A>) -> ()) -> Beacon.CancelableOperation
25 <compiler-generated>    reabstraction thunk helper from @escaping @callee_guaranteed (@guaranteed Foundation.Data?, @guaranteed __C.NSURLResponse?, @guaranteed Swift.Error?) -> () to @escaping @callee_unowned @convention(block) (@unowned __C.NSData?, @unowned __C.NSURLResponse?, @unowned __C.NSError?) -> ()
26 CFNetwork               _CFNetServiceBrowserSearchForServices
27 CFNetwork               __CFHTTPMessageSetResponseProxyURL
28 libdispatch.dylib       __dispatch_call_block_and_release
29 libdispatch.dylib       __dispatch_client_callout
30 libdispatch.dylib       __dispatch_lane_serial_drain
31 libdispatch.dylib       __dispatch_lane_invoke
32 libdispatch.dylib       __dispatch_workloop_worker_thread
33 libsystem_pthread.dylib __pthread_wqthread
34 libsystem_pthread.dylib _start_wqthread
```

**Comment by tiwoc** [comment]
Unfortunately, I wasn't able to reproduce these. Fixing the bugs could be a matter of just dispatching to the main queue at the code locations that the crash reports point to before touching the layout system. Thanks for looking into this!

**Comment by ScottPetit** [comment]
👋 Hey @tiwoc,

Thanks for reporting this. I just wanted to reach out and let you know that we believe we've resolved these crashes in our latest version (2.2.3). If you want to see what else in this release feel free to check out our [Changelog](https://github.com/helpscout/beacon-ios-sdk/blob/master/CHANGELOG.md).
